spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractBinder.java

```diff
@@ -538,31 +538,31 @@ protected RetryTemplate buildRetryTemplateIfRetryEnabled(DefaultBindingPropertie
 
 	/**
 	 * Handles representing any java class as a {@link MimeType}.
+	 *
 	 * @author David Turanski
-	 * @see <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getName"/>
+	 * @author Ilayaperumal Gopinathan
 	 */
 	abstract static class JavaClassMimeTypeConversion {
 
-		public static final MimeType APPLICATION_OCTET_STREAM_MIME_TYPE = MimeType.valueOf
-				(APPLICATION_OCTET_STREAM_VALUE);
-
-		public static final MimeType TEXT_PLAIN_MIME_TYPE = MimeType.valueOf(TEXT_PLAIN_VALUE);
-
 		private static ConcurrentMap<String, MimeType> mimeTypesCache = new ConcurrentHashMap<>();
 
-		static MimeType mimeTypeFromObject(Object obj) {
-			Assert.notNull(obj, "object cannot be null.");
-			if (obj instanceof byte[]) {
-				return APPLICATION_OCTET_STREAM_MIME_TYPE;
+		static MimeType mimeTypeFromObject(Object payload, Object originalContentType) {
```

Comment:
I'm wondering if we can avoid passing an argument here, and move the JSON test outside of it. It seems to imply that the originalContentType plays a bigger role than just preserving the APPLICATION_JSON content type if already set.